### ISSUE TITLE
Checking an optional field incorrectly narrows a union

### ISSUE THREAD

**Issue by lewis6991** [EmmyLuaLs/emmylua-analyzer-rust#1239]
Version: `emmylua_check 0.25.1`, default configuration.

`main.lua`:

```lua
---@class Command
---@field command string

---@class CodeAction
---@field command? Command
---@field disabled? boolean
---@field edit? table

---@param action Command|CodeAction
local function handle(action)
  if type(action.command) == 'string' then
    return
  end
  if action.disabled then
    return
  end
  print(action.edit)
end

handle({ edit = {} })
```

Running `emmylua_check . --warnings-as-errors` reports:

```text
main.lua:17:16
Undefined field `edit`. [undefined-field]
```

Expected: no diagnostic. `CodeAction.disabled` can be nil or false while `edit` exists. Removing the `action.disabled` check makes the checker pass.
